clang/test/Sema/cheri/cheri-cap-ptr-casts.c

```diff
@@ -1,12 +1,26 @@
-// RUN: %cheri_cc1 %s -DALIGN=1 -verify
-// RUN: %cheri_cc1 %s -fsyntax-only -ast-dump 2>/dev/null  | FileCheck %s -check-prefixes AST,HYBRID-AST
+// RUN: %cheri_cc1 %s -DTEST -verify=expected,hybrid
+// RUN: %cheri_cc1 -x c++ %s -DTEST -verify=expected,hybrid
+// RUN: %cheri_purecap_cc1 %s -DTEST -verify=expected,purecap
+// RUN: %cheri_purecap_cc1 -x c++ %s -DTEST -verify=expected,purecap
+// RUN: %cheri_cc1 %s -fsyntax-only -ast-dump 2>/dev/null | FileCheck %s -check-prefixes AST,HYBRID-AST
+// RUN: %cheri_cc1 -x c++ %s -fsyntax-only -ast-dump 2>/dev/null | FileCheck %s -check-prefixes AST,HYBRID-AST
 // RUN: %cheri_purecap_cc1 %s -fsyntax-only -ast-dump 2>/dev/null | FileCheck %s -check-prefixes AST,PURECAP-AST
+// RUN: %cheri_purecap_cc1 -x c++ %s -fsyntax-only -ast-dump 2>/dev/null | FileCheck %s -check-prefixes AST,PURECAP-AST
```

Comment:
```suggestion
// RUN: %cheri_cc1 %s -verify=expected,hybrid,hybrid-c -ast-dump | FileCheck %s -check-prefixes AST,HYBRID-AST
// RUN: %cheri_cc1 -x c++ %s -verify=expected,hybrid,hybrid-cxx -ast-dump | FileCheck %s -check-prefixes AST,HYBRID-AST
// RUN: %cheri_purecap_cc1 %s -verify=expected,purecap -ast-dump | FileCheck %s -check-prefixes AST,PURECAP-AST
// RUN: %cheri_purecap_cc1 -x c++ %s -verify=expected,purecap -ast-dump | FileCheck %s -check-prefixes AST,PURECAP-AST

// XXX: Currently we only get diagnostics for hybrid as DiagnoseCHERIPtr only
// looks through integer pointers (and is only called for C not C++).
// purecap-no-diagnostics
// hybrid-cxx-no-diagnostics
void a() {
  // AST-LABEL: FunctionDecl
  // AST-SAME: a 'void ()'
  unsigned long foo[8];
  // Note: "increases required alignment" is only diagnosed in C mode and
  // err_cheri_ptr_align is also a poor diagnostic message since this is more
  // about creating a pointer to a possibly underaligned capability. Also why is
  // it an error and not a warning...
  ((int * __capability *)foo)[0] = 0;
  // hybrid-c-error-re@-1 {{cast from 'unsigned long *' to 'int * __capability *' increases required alignment from 8 to {{16|32}}}}
  // hybrid-c-note@-2 {{use __builtin_assume_aligned(..., sizeof(void* __capability)) if you know that the source type is sufficiently aligned}}
  // HYBRID-AST:  CStyleCastExpr {{.*}} {{.*}} 'int * __capability *' <BitCast>{{$}}
  // PURECAP-AST: CStyleCastExpr {{.*}} {{.*}} 'int **' <BitCast>{{$}}
  // AST-NEXT:    ImplicitCastExpr {{.*}} {{.*}} 'unsigned long *' <ArrayToPointerDecay> part_of_explicit_cast{{$}}
}
```

This works, but github won't let me attach it sensibly due to unchanged lines inbetween.
Seems like the error message was introduced in https://github.com/CTSRD-CHERI/llvm-project/commit/5eeab18c7345b973d4720aeaf0b62d4fd703b0f8